generic/tests/pktgen_perf.py

```diff
@@ -0,0 +1,197 @@
+                runner = session.cmd
+                pkt_cate_r = run_test(session_serial, runner, remote_path,
+                                      pktgen_ip, dsc, pktgen_interface,
+                                      run_threads, size, timeout)
+            elif pkt_cate == "rx":
+                error_context.context("test guest rx pps performance",
+                                      logging.info)
+                host_bridge = params.get("netdst", "switch")
+                host_nic = utils_net.Interface(host_bridge)
+                pktgen_ip = host_nic.get_ip()
+                dsc = vm.wait_for_get_address(0, timeout=5)
+                pktgen_interface = vm.get_ifname(0)
+                runner = process.system_output
+                pkt_cate_r = run_test(session_serial, runner, remote_path,
+                                      pktgen_ip, dsc, pktgen_interface,
+                                      run_threads, size, timeout)
+            line = ""
+            line += "%s|" % format_result(size)
+            line += "%s" % format_result(pkt_cate_r)
+            result_file.write(("%s\n" % line))
+
+    error_context.context("Verify Host and guest kernel no error\
+                           and call trace", logging.info)
+    vm.verify_kernel_crash()
+    utils_misc.verify_dmesg()
+    session_serial.close()
+    session.close()
+
+
+def run_test(session_serial, runner, remote_path, pktgen_ip, dsc,
+             interface, run_threads, size, timeout):
+    """
+    Run pktgen_perf script on remote and gather packet numbers/time and
+    calculate mpps.
+
+    :param session_serial: session serial for vm.
+    :param runner: connection for vm or host.
+    :param remote_path: pktgen_perf script path.
+    :param pktgen_ip: ip address which pktgen script was running.
+    :param dsc: dsc mac or dsc ip pass to pktgen_perf script.
+    :param interface: device name pass to pktgen_perf script.
+    :param run_threads: the numbers pktgen threads.
+    :param size: packet size pass to pktgen_perf script.
+    :param timeout: test run time.
+    """
+    exec_cmd = "%s %s %s %s %s" % (remote_path, dsc, interface,
+                                   run_threads, size)
+    packets = "cat /sys/class/net/%s/statistics/tx_packets" % interface
+    logging.info("Start pktgen test by cmd '%s'" % exec_cmd)
+    try:
+        packet_b = runner(packets)
+        runner(exec_cmd, timeout)
+    except (aexpect.ShellTimeoutError, process.CmdError), err:
+        if isinstance(err, aexpect.ShellTimeoutError):
+            # when pktgen script is running on guest, it's damaged,
+            # guest could not response by ssh, so uses serial instead.
+            packet_a = session_serial.cmd(packets, timeout)
+            kill_cmd = "kill -9 `ps -C pktgen_perf.sh -o pid=`"
+            session_serial.cmd(kill_cmd)
+            session_serial.cmd("ping %s -c 5" % pktgen_ip,
+                               ignore_all_errors=True)
```

Comment:
You do not need to check `ping` success or not?

Reply:
I think no need to check it, use ping to let guest's network re-active, otherwise, guest could not be connected.